### PR TITLE
optimizer: Do inner join order optimization and add test cases.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @qiuyesuifeng PTAL

File: optimizer/plan/planbuilder_join.go [hunk 1]
```diff
@@ -383,23 +391,160 @@ func (p *joinPath) removeIndexDepCycle(origin *joinPath) {
...
+		availablePaths = append(availablePaths, next)
+		for path := range pathMap {
+			if path.idxDeps != nil {
+				delete(path.idxDeps, next)
+			}
+			if path.neighbors != nil {
+				delete(path.neighbors, next)
+			}
+		}
+		p.reattach(pathMap, availablePaths)
```
Comment (ngaut) [inline optimizer/plan/planbuilder_join.go]: s/reattach/reAttach/g

File: optimizer/plan/plan_test.go [hunk 2]
```diff
@@ -346,6 +346,38 @@ func (s *testPlanSuite) TestSplitWhere(c *C) {
...
+		expr    string
+		notNull bool
+	}{
+		{"a = 1", true},
+		{"a != 100 and a > 0", true},
+		{"a is null", false},
+		{"a is not true", false},
+		{"a is not false", false},
+		{"a != 0 and a is not false", true},
+		{"a is not true", false},
```
Comment (qiuyesuifeng) [inline optimizer/plan/plan_test.go]: Duplicated?

File: optimizer/plan/planbuilder_join.go [hunk 3]
```diff
@@ -14,25 +14,26 @@
...
 	"github.com/pingcap/tidb/parser/opcode"
 )
 
-// Equiv represents an equivalent join condition, like "t1.c1 = t2.c1".
-type Equiv struct {
-	Left     *ast.ResultField
-	LeftIdx  bool
-	Right    *ast.ResultField
-	RightIdx bool
+// equalCondition represents an equivalent join condition, like "t1.c1 = t2.c1".
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: equalCond

File: optimizer/plan/planbuilder_join.go [hunk 4]
```diff
@@ -72,7 +72,7 @@ type joinPath struct {
 	parent     *joinPath
 	filterRate float64
 	conditions []ast.ExprNode
-	equivs     []*Equiv
+	equivs     []*equalCond
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: eqConditions?

File: optimizer/plan/planbuilder_join.go [hunk 5]
```diff
@@ -214,20 +217,26 @@ func (p *joinPath) containsTable(table *ast.TableName) bool {
 	return p.outer.containsTable(table) || p.inner.containsTable(table)
 }
 
-// attachEquiv tries to attach a Equiv deep into a table path if applicable.
-func (p *joinPath) attachEquiv(equiv *Equiv, availablePaths []*joinPath) (attached bool) {
+// attachEqualCond tries to attach a equalCond deep into a table path if applicable.
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: an equalCond?

File: optimizer/plan/planbuilder_join.go [hunk 6]
```diff
@@ -383,23 +392,162 @@ func (p *joinPath) removeIndexDepCycle(origin *joinPath) {
...
+	p.inners = ordered
 }
 
-func (p *joinPath) partialScore() float64 {
-	return 1 / p.filterRate
+// reattach is called by inner joinPath to retry attach conditions to inner paths
+// after an inner path has been added to available paths.
+func (p *joinPath) reattach(pathMap map[*joinPath]bool, availablePaths []*joinPath) {
+	if len(p.conditions) != 0 {
+		conMap := map[ast.ExprNode]bool{}
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Seems slice is enough.

File: optimizer/plan/planbuilder_join.go [hunk 7]
```diff
@@ -383,23 +392,162 @@ func (p *joinPath) removeIndexDepCycle(origin *joinPath) {
...
+		}
+		cans = append(cans, t)
+	}
+	return cans
+}
+
+func (p *joinPath) nextIndexPath(candidates []*joinPath) *joinPath {
+	var indexPaths []*joinPath
+	for _, can := range candidates {
+		neighborIsAvailable := len(can.neighbors) == 0 && can.neighborCount > 0
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Please add some comments.

File: optimizer/plan/planbuilder_join.go [hunk 8]
```diff
@@ -479,7 +628,10 @@ func (n *nullRejectFinder) Enter(in ast.Node) (ast.Node, bool) {
 func (n *nullRejectFinder) Leave(in ast.Node) (ast.Node, bool) {
 	switch x := in.(type) {
 	case *ast.ColumnNameExpr:
-		n.nullRejectTalbes[x.Refer.TableName] = true
+		if x.Refer == nil {
+			panic(x.Name.Name)
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Why here panic?

File: optimizer/plan/planbuilder_join.go [hunk 9]
```diff
@@ -383,23 +392,161 @@ func (p *joinPath) removeIndexDepCycle(origin *joinPath) {
...
+	return cans
+}
+
+func (p *joinPath) nextIndexPath(candidates []*joinPath) *joinPath {
+	var indexPaths []*joinPath
+	for _, can := range candidates {
+		// Since we may not have equal conditions attached on the path, we
+		// need to check neighborCount and idxDepCount to see if this path
+		// can be joined with index.
+		neighborIsAvailable := len(can.neighbors) == 0 && can.neighborCount > 0
```
Comment (shenli) [inline optimizer/plan/planbuilder_join.go]: len(can.neighbors) < can.neighborCount?

File: optimizer/plan/planbuilder_join.go [hunk 10]
```diff
@@ -383,23 +392,161 @@ func (p *joinPath) removeIndexDepCycle(origin *joinPath) {
...
+	var innerStrs []string
+	for _, in := range p.inners {
+		innerStrs = append(innerStrs, in.String())
 	}
+	return "inner{" + strings.Join(innerStrs, ",") + "}"
 }
 
-func (p *joinPath) totalScore() float64 {
-	return float64(p.revDepCount+1) / p.totalFilterRate
+func (p *joinPath) optimizeJoinOrder(availablePaths []*joinPath) {
```
Comment (shenli) [inline optimizer/plan/planbuilder_join.go]: Seems availablePaths is useless.

File: optimizer/plan/planbuilder_join.go [hunk 11]
```diff
@@ -383,23 +392,161 @@ func (p *joinPath) removeIndexDepCycle(origin *joinPath) {
...
+		idxDepIsAvailable := can.idxDepCount > 0
+		if can.hasOuterIdxEqualCond() || neighborIsAvailable || idxDepIsAvailable {
+			indexPaths = append(indexPaths, can)
+		}
+	}
+	if len(indexPaths) == 0 {
+		return nil
+	}
+	var best *joinPath
+	for _, path := range indexPaths {
```
Comment (shenli) [inline optimizer/plan/planbuilder_join.go]: Merge this loop into the above one.

File: optimizer/plan/planbuilder_join.go (hunk 10 above)
Comment (coocood) [inline optimizer/plan/planbuilder_join.go]: It is used in `reattach`

**Comment by shenli** [comment]
LGTM

**Pull request merged**